### ISSUE TITLE
Summary of GSOC2019 calls

### ISSUE THREAD

**Issue by avinashbarnwal** [avinashbarnwal/xgboost#1]
**21MAY2019**
Relevant Information - 
1. What should be xgb dmatrix for Interval regression ?
Ans :- Create two column for all the times
2-column matrix representation of outputs=labels.
 1. un-censored output, event is obserrved, e.g. y_i = 5. (5, 5)
> survival::Surv(10, 10, type="interval2")
[1] 10
2. left-censored output, event not observed, but we know it happened some time before t_i=5. y_i = (-Inf = \underline y_i, 5=t_i=\overline y_i)
> survival::Surv(-Inf, 10, type="interval2")
[1] 10-
3. Right-censored output, event not observed , but we know it happened some time after t_i =5. y_i = (5,inf)
> survival::Surv(5, Inf, type="interval2")
[1] 5+
4. Interval-censored output, event not observed, but we know it happened some time in y_i =  (5,10) 
> survival::Surv(5, 10, type="interval2")
[1] [5, 10]

There is no need for the survival object. This is more of a legacy.

2. How to handle sigma in AFT?
Ans:- Treat as a hyperparameter.

3. What is the dimension of the predicted value of interval regression?
Ans:- It is always one real value, not interval.

Relevant Document
https://github.com/tdhock/aft-poster/blob/master/HOCKING-AFT.pdf 
http://members.cbio.mines-paristech.fr/~thocking/survival.pdf
https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-9-269

Please Check.
@tdhock, @hcho3

**Comment by chyunsu3** [comment]
**May 28, 2019 Update**

1. Done: re-created the loss charts from Toby’s paper to show that loss functions are reasonable
2. Done: submitted RFC to public XGBoost repo
3. In-Progress: Python Proof-of-Concept for end-to-end gradient boosting with survival loss
4. In-Progress: Derive formulas for AFT

@avinashbarnwal @tdhock

**Comment by tdhock** [comment]
hi @avinashbarnwal  can you please post the source code and rendered images for the loss charts you created?

**Comment by chyunsu3** [comment]
@tdhock See https://github.com/avinashbarnwal/gsoc/tree/master/AFT/R

**Comment by tdhock** [comment]
looks like a good start but the plot is incorrect for uncensored data. it should look like the square loss around the label for the normal distribution. also you should try creating a facetted ggplot, with one panel per censoring type. for inspiration here is the code I used for the 1-page AFT poster, https://github.com/tdhock/aft-poster/blob/master/figure-loss.R

**Comment by avinashbarnwal** [comment]
Thanks, Prof. @tdhock  I am looking into it.

**Comment by tdhock** [comment]
looks better @avinashbarnwal ! glad to see that you got the facetted ggplots working.

however it looks like there is a problem with your computation of the logistic loss for the uncensored output -- it should be minimal at the label.

- the x axis title says log scale but the ticks indicate linear scale (i.e the distance between 2.5 and 5.0 is the same as between 5.0 and 7.5)
- the x axis title also says that you are plotting the predicted survival time in days but I don't think that is right -- aren't you actually plotting the predicted center of the gaussian/logistic = log(predicted time) ?
- the legend says logLogistic and logNormal but you may want to change that to normal / logistic for simplicity -- it is usually easier to deal with the real valued labels rather than the positive valued labels. technically if t_i>0 is the survival time and y_i = log(t_i) \in R is the real-valued log(survival time) then the model assumes t_i  ~ lognormal( f(x_i), \sigma^2) which is the same thing as y_i ~ normal( f(x_i), \sigma^2)
- please increase the number of grid points at which you evaluate the loss functions (the curves are pretty choppy)
- please increase the size of the interval censored output, so we can see the loss function approach zero and flatten out at the center of the limits.

**Comment by tdhock** [comment]
also for next week's homework please check your formulas for the gradient in the overleaf. add a row of plots to the figure that shows the loss functions. use `facet_grid(fun ~ type)` -- rows for different functions (loss, gradient, hessian)
1. first row should be loss function
2. second row should be gradient

**Comment by avinashbarnwal** [comment]
Prof. @tdhock, @hcho3 

I have made the loss function with all the changes required.
Here is the link for the Plot - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/loss_aft.png

I have changed the code as well. 
Code Link - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/assigment1.R

Prof. @tdhock  Please check the formula for the Normal - Uncensored part. I have used different formula compared to given in this https://github.com/avinashbarnwal/GSOC-2019/blob/master/paper/HOCKING-AFT.pdf, rather than this I have used the formula given in this document. (http://home.iitk.ac.in/~kundu/paper146.pdf).

**Comment by tdhock** [comment]
looks better Avinash. but why doesn't the uncensored loss go to zero? (it
should...)

about the normal - censored loss, you should double check your work by
using the normal CDF  (pnorm in R)

**Comment by avinashbarnwal** [comment]
Prof. @tdhock, The uncensored 
Loss function Normal- -log(1/(t.lower * sigma * sqrt(2 * pi)) * exp((log(t.lower/y.hat))**2/(-2 * sigma * sigma))) 

We have a constant term -log(1/(t.lower * sigma * sqrt(2*pi)) irrespective of y.hat. This makes it non-zero, this is based on my thinking.

Similarly for Logistic.

Sorry, I meant Normal Uncensored formula in the above.

Normal Uncensored Old Formula - 
-log(1/(y.hat * sigma * sqrt(2 * pi)) * exp((log(t.lower/y.hat))**2/(-2 * sigma * sigma))) 


Normal Uncensored New Formula - 
-log(1/(t.lower * sigma * sqrt(2 * pi)) * exp((log(t.lower/y.hat))**2/(-2 * sigma * sigma)))

**Comment by tdhock** [comment]
right but we usually subtract away the constant terms. if you do that you
should recover the square loss which is 0 at the min.

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock, @hcho3 

Please find the updated document for AFT(https://github.com/avinashbarnwal/GSOC-2019/blob/master/doc/Accelerated_Failure_Time.pdf). I think I have made mistake here as I am taking gradient and hessian with respect to y.hat, not X\beta.

For least square loss function-gradient and hessian with respect to y.hat and X\beta are same. But it would start mattering when we have link functions involved. Similarly, for classification, we take gradient and hessian with respect to  X\beta, not y.hat. 

For reference, please check doc - https://cran.r-project.org/web/packages/gbm/vignettes/gbm.pdf
Bernoulli part. 

This leads to more decision on notations as in survival document we are using f for pdf and in the above document, we have f(x_i) for X\beta. 

Please let me know your thoughts.

**Comment by tdhock** [comment]
I dont think there is an issue because the link function is always the identity with the normal and logistic models. again we should be able to use the formulas in the survival manual.

**Comment by avinashbarnwal** [comment]
Prof. @tdhock, I think we need to calculate gradient with respect to log(y.hat) not y.hat as log(y.hat) = X\beta. Please let me know.

**Comment by tdhock** [comment]
section 6.8 of survival manual gives derivatives with respect to eta, which is the real-valued prediction.

**Comment by tdhock** [comment]
in your notation you use log(y.hat) for the real valued prediction, so that is the same as eta from the survival manual.

in your PDF please do not use x\beta as that is only valid for linear models

**Comment by avinashbarnwal** [comment]
Prof. @tdhock, Thanks. I will use eta for log(y.hat) and take gradient and Hessian based on that. In my document, I was taking gradient and Hessian based on y.hat, not on eta. I will make the correction.

**Comment by avinashbarnwal** [comment]
Prof. @tdhock, @hcho3, I have created the document and plot for loss, negative gradient, and hessian.    

Document - https://github.com/avinashbarnwal/GSOC-2019/blob/master/doc/Accelerated_Failure_Time.pdf
Plot - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/loss_grad_hess_aft.png
Code - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/combined_assignment.R

Please let me know your thoughts.

**Comment by tdhock** [comment]
the plot looks good except for the hessian for the interval censored outputs for small predictions -- it seems to be too large. can you please double check? it should look like right censored outputs (i.e. with finite lower limit) on the left side of the plot

**Comment by avinashbarnwal** [comment]
Thanks, Prof. @tdhock.  Please check it again.

**Comment by avinashbarnwal** [comment]
Thanks, Prof. @tdhock. I have changed the code and plot both. I had one sign in the formula wrong. Please recheck the plot (https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/loss_grad_hess_aft.png). Do you think Interval hessian is correct for Normal distribution?

**Comment by avinashbarnwal** [comment]
Hi @hcho3 and Prof. @tdhock,

Please find the python implementation of gradient boosting for AFT.

Code  - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/py/gb_aft.ipynb
Plot    - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/py/Nesterov_False.png

and let me know your thoughts.

**Comment by chyunsu3** [comment]
We'll need to check the POC implementation, since the training loss should trend down, not up.
Also, let's plot `mean(abs(log(Y) - eta))`, since the predicted score `eta` should try to match `log(Y)`.

**Comment by avinashbarnwal** [comment]
Hi @hcho3 and Prof. @tdhock,

As discussed with @hcho3, I am working on documentation for binomial loss and POC for the same.

**Comment by avinashbarnwal** [comment]
Hi @hcho3 and Prof. @tdhock,

Please find the document for binomial loss and code below:-

Document -  https://github.com/avinashbarnwal/GSOC-2019/blob/master/doc/Binomial_Loss.pdf.
Code          - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/binomial_loss.R.
Plot            - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/binomial_loss.png.

**Comment by tdhock** [comment]
there is still something wrong for the interval hessian for small predicted values on https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/loss_grad_hess_aft.png -- it should be constant as the prediction goes to zero, log(prediction) goes to -Inf

**Comment by tdhock** [comment]
binomial loss looks reasonable, except for the x axis label.

please (1) use facet_grid and (2) use more grid points and (3) maybe have different columns or colors for different labels

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock  and @hcho3 ,

Please check the R plots and Python Plots again -

R Plots - 
1. AFT - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/loss_grad_hess_aft.png
2. Binomial Loss - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/binomial_loss.png

Python Plots-
1. Log loss, Data Type - Mixed - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/py/Nesterov_False_Loss_LogLoss_Data_Mixed.png
2. Mae, Data Type      - Mixed - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/py/Nesterov_False_Loss_Mae_Data_Mixed.png
3. Log loss, Data Type - Uncen - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/py/Nesterov_False_Loss_Logloss_Data_Uncensored.png
4. Mae, Data Type      - Uncen - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/py/Nesterov_False_Loss_Mae_Data_Uncensored.png

**Comment by tdhock** [comment]
aft hessian looks good now but binomial hessian has a problem: should be
zero as prediction goes to Inf

**Comment by tdhock** [comment]
paper with an example of a real-world problem with only left, right, and interval censored labels: http://proceedings.mlr.press/v28/hocking13.html

**Comment by tdhock** [comment]
discussed:
1. compute loss via dnorm or log/exp? in C++ normal distn functions are provided in std lib, but logistic dist functions are not. logistic provided in boost but we don't want to depend on that.
2. check work on binomial loss via duplicate data trick.
3. how to specify binomial labels? most standard way would be two-column matrix of #success and #possible as in https://en.wikipedia.org/wiki/Binomial_distribution
4. how to specify censored labels? two-column matrix, no event indicator needed.

**Comment by avinashbarnwal** [comment]
Hi, Prof @tdhock and @hcho3,

 Please check the updated plot and code for binomial loss.

Plot - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/binomial_loss.png
Code - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/binomial_loss.R
Doc - https://github.com/avinashbarnwal/GSOC-2019/blob/master/doc/Binomial_Loss.pdf

**Comment by tdhock** [comment]
that looks more reasonable

suggestions: (1) more grid points, (2) no need to use aes(color) because
you already put the different functions in different panels, (3) add more
columns for different labels, e.g. y=5, n=10; y=0, n=10;, y=2, n=10

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock ,

Please find the updated binomial loss:-
**R**
Binomial Plot - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/binomial_loss.png.
AFT Plot - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/loss_grad_hess_aft.png.

**Py**
**Updated Plots**

LogLoss - Mixed - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/py/Nesterov_False_Loss_LogLoss_Data_Mixed.png
Mae        - Mixed - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/py/Nesterov_False_Loss_Mae_Data_Uncensored.png
LogLoss - Uncensored - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/py/Nesterov_False_Loss_Logloss_Data_Uncensored.png
Mae        - Uncensored - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/py/Nesterov_False_Loss_Mae_Data_Uncensored.png

**C++**
Functions for Pdf and cdf of Normal and Logistic Distribution

https://github.com/avinashbarnwal/GSOC-2019/tree/master/AFT/C%2B%2B

I have tested with x=0 for each distributions and it was giving correct results compared to R inbuilt functions.

Thanks, @hcho3 for changing the hessian formula for interval data.

**Comment by chyunsu3** [comment]
@avinashbarnwal The AFT loss plots now look good. I'll take a look at the C++ code soon.

**Comment by avinashbarnwal** [comment]
Hi  Prof. @tdhock and @hcho3 

I have been working on implementing loss, negative gradient and hessian functions for AFT in C++ and test it through plots on Python.

For details- Please find the link.
https://github.com/avinashbarnwal/GSOC-2019/tree/master/AFT/C%2B%2B

Please let me know your thoughts.

**Comment by chyunsu3** [comment]
@avinashbarnwal I modified your notebook slightly to plot the uncensored AFT loss with yhat in the X axis:
![download](https://user-images.githubusercontent.com/2532981/61324681-de373000-a7c7-11e9-85fb-6ad6cc458c90.png)

The curve for the normal should not cliff off like that.

**Comment by chyunsu3** [comment]
In general, let's try to reproduce the R plots using C++ code. That way, we have assurance that C++ code does what R code does.

**Comment by tdhock** [comment]
right, the loss function should be convex (for a fixed scale parameter)

**Comment by tdhock** [comment]
hi there @avinashbarnwal is the R learning algo for AFT losses working yet?

When it is I would suggest modifying https://github.com/tdhock/neuroblastoma-data/blob/master/iregnet.R which I wrote for benchmarking a different learning algorithm for censored outputs, on 33 different labeled data sets.

**Comment by chyunsu3** [comment]
@tdhock Thanks for the link. The datasets should come in handy.

**Comment by avinashbarnwal** [comment]
Prof. @tdhock ,

For both AFT and binomial loss, R learning algorithms(loss, negative gradient and hessian) are working.
I am currently writing C++ functions for binomial loss. Soon we will have everything in the xgboost package.

**Comment by tdhock** [comment]
ok where is the fork/branch with the AFT models? can I install it?

Here is a new benchmark script https://github.com/tdhock/neuroblastoma-data/blob/master/xgboost.R#L41 

The benchmark includes 33 different data sets, each with several train/test splits (designated by fold ID numbers in folds.csv file).

To run the code please fork then clone that repo, and then modify the "xgboost.R" script so xgboost works on these learning problems. (run the script / start R in the neuroblastoma-data directory)

The script also runs baselines penaltyLearning::IntervalRegressionCV (penaltyLearning.scale1) and constant (always predict 0).

**Comment by chyunsu3** [comment]
@avinashbarnwal 

> R learning algorithms(loss, negative gradient and hessian) are working

I don't think you have boosting PoC in R? We do have boosting PoC in Python, however.

**Comment by avinashbarnwal** [comment]
Prof. @tdhock ,

We haven't add the AFT loss function in xgboost package yet. We are testing it first in C++ and then I will add to the package. Please let me know if you want proof of concept in R.

**Comment by chyunsu3** [comment]
@avinashbarnwal For now, can we run Python boosting PoC on the 33 datasets?

**Comment by tdhock** [comment]
no problem if it is not ready yet but you may consider using those data sets to test /debug the algo.

if you don't use that R script then you need to know that
1. input/feature matrix is in data/DATA_NAME/inputs.csv.xz -- some features are missing (NA) or infinite in some rows -- you can ignore those features/columns.
2. output/label matrix is in data/DATA_NAME/outputs.csv.xz
3. would be good to compare to another learner, at least a baseline constant prediction, would be better to compare against regularized linear model / computing test AUC as in my R script.
4. suggested cross-validation folds are given in folds.csv files, e.g. 
```r
> folds.csv.vec <- Sys.glob("data/*/cv/*/folds.csv")
> folds.csv.vec
 [1] "data/ATAC_JV_adipose/cv/equal_labels/folds.csv"         
 [2] "data/CTCF_TDH_ENCODE/cv/equal_labels/folds.csv"         
 [3] "data/H3K27ac-H3K4me3_TDHAM_BP/cv/equal_labels/folds.csv"
 [4] "data/H3K27ac_TDH_some/cv/equal_labels/folds.csv"        
 [5] "data/H3K27me3_RL_cancer/cv/equal_labels/folds.csv"      
 [6] "data/H3K27me3_TDH_some/cv/equal_labels/folds.csv"       
 [7] "data/H3K36me3_AM_immune/cv/equal_labels/folds.csv"      
 [8] "data/H3K36me3_TDH_ENCODE/cv/equal_labels/folds.csv"     
 [9] "data/H3K36me3_TDH_immune/cv/equal_labels/folds.csv"     
[10] "data/H3K36me3_TDH_other/cv/equal_labels/folds.csv"      
[11] "data/H3K4me1_TDH_BP/cv/equal_labels/folds.csv"          
[12] "data/H3K4me3_PGP_immune/cv/equal_labels/folds.csv"      
[13] "data/H3K4me3_TDH_ENCODE/cv/equal_labels/folds.csv"      
[14] "data/H3K4me3_TDH_immune/cv/equal_labels/folds.csv"      
[15] "data/H3K4me3_TDH_other/cv/equal_labels/folds.csv"       
[16] "data/H3K4me3_XJ_immune/cv/equal_labels/folds.csv"       
[17] "data/H3K9me3_TDH_BP/cv/equal_labels/folds.csv"          
[18] "data/detailed/cv/R-3.6.0-chrom/folds.csv"               
[19] "data/detailed/cv/R-3.6.0-profileID/folds.csv"           
[20] "data/detailed/cv/R-3.6.0-profileSize/folds.csv"         
[21] "data/detailed/cv/R-3.6.0-sequenceID/folds.csv"          
[22] "data/detailed/cv/chrom/folds.csv"                       
[23] "data/detailed/cv/profileID/folds.csv"                   
[24] "data/detailed/cv/profileSize/folds.csv"                 
[25] "data/detailed/cv/sequenceID/folds.csv"                  
[26] "data/systematic/cv/R-3.6.0-chrom/folds.csv"             
[27] "data/systematic/cv/R-3.6.0-profileID/folds.csv"         
[28] "data/systematic/cv/R-3.6.0-profileSize/folds.csv"       
[29] "data/systematic/cv/R-3.6.0-sequenceID/folds.csv"        
[30] "data/systematic/cv/chrom/folds.csv"                     
[31] "data/systematic/cv/profileID/folds.csv"                 
[32] "data/systematic/cv/profileSize/folds.csv"               
[33] "data/systematic/cv/sequenceID/folds.csv"                
> 
```

**Comment by avinashbarnwal** [comment]
Thanks. Prof. @tdhock and @hcho3 I will show the results for given datasets.

**Comment by tdhock** [comment]
Also if we have time we should consider supporting the Extreme Value distribution (in addition to logistic and normal), since in stats people often like to use Weibull/Exponential models, and they are special cases of the EV dist.

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock and @hcho3,

I have added the C++ code for binomial loss and notebook for validating the same.
Links
C++  - https://github.com/avinashbarnwal/GSOC-2019/tree/master/BinomialLoss/C%2B%2B
Notebook  - https://github.com/avinashbarnwal/GSOC-2019/blob/master/BinomialLoss/Python%20Notebook/Visualizing%20distributions.ipynb

I am in the process of testing the AFT for 33 datasets.
Link - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/test/data/neuroblastoma-data-master/src/notebook/001_data_massage.ipynb.

For Extreme Value Distribution, I will look into it as we have Extreme Value Distribution in the survival document shared by you. I will change the R, Python and C++ Plots accordingly.

**Comment by chyunsu3** [comment]
@avinashbarnwal I left a comment about the binomial loss implementation: avinashbarnwal/GSOC-2019#5. We do not want to use `n!/ ((n-r)! r!)`  because 1) it has many redundant computation and 2) it may cause overflows.

**Comment by chyunsu3** [comment]
@tdhock I modified @avinashbarnwal's notebook a bit to generate a combined plot with the C++ AFT loss:

![combined_plot_python](https://user-images.githubusercontent.com/2532981/61751089-012f8a00-ad5c-11e9-80a1-5daa4f738df7.png)

So we're pretty confident that his AFT implementation is correct.

**Comment by tdhock** [comment]
that looks reasonable

**Comment by chyunsu3** [comment]
I added some stub code in the XGBoost codebase, so that we can store `y_lower` and `y_upper` as part of the data matrix (DMatrix) object: https://github.com/avinashbarnwal/xgboost/commits/survival_analysis

@avinashbarnwal will be working to graft the AFT loss into XGBoost.

Usage:

```python
import xgboost

# Create data matrix
dtrain = xgboost.DMatrix('./demo/data/agaricus.txt.train')
# Set y_lower
dtrain.set_float_info('label_lower_bound', np.random.rand(6513))
# Set y_upper
dtrain.set_float_info('label_upper_bound', np.random.rand(6513))
# Get y_lower and y_upper
print(dtrain.get_float_info('label_lower_bound'))
print(dtrain.get_float_info('label_upper_bound'))
```

We'll be adding more user-friendly interfaces later on.

**Comment by avinashbarnwal** [comment]
Hi @hcho3  and Prof. @tdhock,

We have been able to test the xgboost for aft for dataset - detailed, systematic and simulated datasets. Detailed and Systematic datasets are mentioned above and Simulated dataset is generated by me. 

We have tested for aft normal.

Please find the plots here - https://github.com/avinashbarnwal/GSOC-2019/tree/master/AFT/test/data/neuroblastoma-data-master/src/notebook/xgboost%20plots.

But this is not working for first few datasets.

Link for notebook - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/test/data/neuroblastoma-data-master/src/notebook/002_xgboost_test.ipynb

Please let me know your thoughts.

**Comment by chyunsu3** [comment]
@avinashbarnwal Nicely done. This is first time AFT loss is implemented in XGBoost.

> But this is not working for first few datasets

Can you be more specific? Why do you think XGBoost is not “working”?

**Comment by avinashbarnwal** [comment]
I think we have not put the overflow in the formula as we have here https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/py/_aft_loss.py in the denominator.

We are getting Nans in the train and test score like this:-
[0]	train-normal,1.0:nan	test-normal,1.0:nan

**Comment by chyunsu3** [comment]
@avinashbarnwal I’m still curious about your line of reasoning here. What indicators or symptoms do you see that you think indicate deficiencies in the current implementation?

Once we know the indicators (“bad smells”), then we can discuss possible diagnosis for that

**Comment by avinashbarnwal** [comment]
I have been trying to print the first order and second order  https://github.com/avinashbarnwal/GSOC-2019/blob/master/xgboost/src/objective/aft_obj.cc. But it is not printing.

**Comment by chyunsu3** [comment]
Use `LOG(CONSOLE)` in place of `std::cout`.

**Comment by avinashbarnwal** [comment]
Hi,

Please find the first order and second order for each iteration.
![log](https://user-images.githubusercontent.com/6061417/62347176-d4e7dc00-b4c6-11e9-86ba-3097ae7c5b30.png)
LOG(CONSOLE) was not giving the right answer. I have attached the std::cout.

**Comment by chyunsu3** [comment]
@avinashbarnwal Can you add a space between two numbers?

**Comment by avinashbarnwal** [comment]
Please find with spaces.
![Log](https://user-images.githubusercontent.com/6061417/62372509-a9d0ad00-b505-11e9-9265-2f20b97b30c0.png)

**Comment by tdhock** [comment]
I don't see what is the problem with the plots.

train error decreasing and validation error going up after some iterations
= looks ok

I would suggest computing test error in K-fold CV, and compare with test
error of baselines which have been computed in the neuroblastoma-data repo

**Comment by chyunsu3** [comment]
@tdhock The issue is that we have infinities in the first and second order gradients.

**Comment by tdhock** [comment]
please double check your C++ code with the R code

is it specific to one loss function?

**Comment by avinashbarnwal** [comment]
In C++ code, we have denominators in the first-order gradient going to 0. We have a problem similar to this. https://github.com/dmlc/xgboost/issues/2955. We can use a similar idea to prevent the overflow.

**Comment by avinashbarnwal** [comment]
Hi @hcho3 and Prof. @tdhock,

I have added the plot for first data -  "data/ATAC_JV_adipose/cv/equal_labels/folds.csv" 

Plots - https://github.com/avinashbarnwal/GSOC-2019/tree/master/AFT/test/data/neuroblastoma-data-master/src/notebook/xgboost%20plots/ATAC_JV_adipose

I have changed the overflow problem as mentioned above. Please find the updated code.
Code - https://github.com/avinashbarnwal/xgboost/blob/survival_analysis/src/objective/aft_obj.cc

Please let me know what you think?

Below is the log-
Now we have the first order and second-order gradient as very small values.
![Log1](https://user-images.githubusercontent.com/6061417/62406052-0b7b3080-b573-11e9-9cc7-a08f8378596d.png)

**Comment by tdhock** [comment]
plots look good

this is the first time I look at your C++ implementation and I think it
would be better to treat left/right as special cases of interval -- do you
think that would be possible? it would reduce duplication in the code...

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock,

I fear, I won't be able to combine left/right-censored to the interval as we have a different formulation for each censored data types.

Hi @hcho3,

As discussed, Please find the negative gradients plots - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/Python%20Notebook/Visualizing%20distributions.ipynb and check the cell number 11. I have changed the range from
#x_vals   = np.linspace(np.log(2**1), np.log(2**15), 200)
to 
x_vals   = np.linspace(np.log(2**1), 1000, 200)

I see that it is going to nan very quickly.

**Comment by tdhock** [comment]
you are right that "we have a different formulation for each censored data types" but the left and right can be expressed as special cases of interval, which would simplify the code.

e.g. 
```
 grad  = (pdf_u-pdf_l)/(sigma*std::max(cdf_u-cdf_l,eps)); //interval
    grad = pdf/(sigma*cdf); //left
// could be instead:
if(y_lower== -INFINITY) {pdf_l=0;cdf_l=0}
```

Also I would suggest simplifying aft_obj.cc by creating a class for each different AFT distribution. This would reduce duplication between the _interval and _uncensored functions e.g.

```
// right now:
grad_interval(){
      case AFTNoiseDistribution::kNormal:
      pdf_u  = common::aft::dnorm(z_u,0,1);
      pdf_l  = common::aft::dnorm(z_l,0,1);
      cdf_u  = common::aft::pnorm(z_u,0,1);
      cdf_l  = common::aft::pnorm(z_l,0,1);
      break;
grad_uncensored(){
     case AFTNoiseDistribution::kNormal:
      pdf  = common::aft::dnorm(z,0,1);
      _grad = grad_norm(z,0,1);
      break;
//suggestion:
class AFTLoss: // base class
  double grad_uncensored(...){
    pdf = pnorm(z, 0, 1);
    _grad = compute_gradient(z, 0, 1); // compute_gradient is defined for each sub-class
    grad = _grad/(sigma*pdf);
class AFTNormal: AFTLoss
  dnorm = common::aft::dnorm;
  pnorm = common::aft::pnorm;
  double compute_gradient(...){
    // same as your current grad_norm function
  }
class AFTLogistic: AFTLoss
  dnorm = common::aft::dlogistic
  double compute_gradient(...){
 // etc
```
by using the base AFTLoss class you reduce duplication

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock,

Thanks for explaining more. I have got it. I will make the changes accordingly.

**Comment by chyunsu3** [comment]
@tdhock Given that GSOC end is coming, can we postpone code re-factoring until after we can get AFT and binomial loss working? Currently @avinashbarnwal needs to finish putting in the binomial loss into XGBoost.

**Comment by chyunsu3** [comment]
@avinashbarnwal @tdhock And ideally, I'd like to see pull request for AFT at [dmlc/xgboost](https://github.com/dmlc/xgboost) pretty soon. As part of the review process, we can re-organize the AFT loss.

**Comment by avinashbarnwal** [comment]
Hi @hcho3 and Prof. @tdhock,

I was testing the algorithm with different eps for different datasets like first and second dataset mentioned above. 

eps = 1e-12f gave the best result. I have kept that as final eps to be used.

**Comment by tdhock** [comment]
I wouldn't suggest to compromise on code quality just because the end of GSOC is near. I think at least one PR (with good quality code) on the main xgboost repo should constitude a successful GSOC project. (IMO no problem if there is not time for the binomial loss)

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock and @hcho3,

I am doing my best to make changes suggested by Prof. @tdhock. As suggested by @hcho3, we won't be able to test the class in python through c++. We can directly test with datasets and manual testing.

I am going to maintain whatever we have in survival_analysis branch in xgboost and push all the changes to the new branch I am going to create named as "survival_analysis1".

I have also added Weibull distribution in R plot. Please find the attached.
https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/loss_grad_hess_aft.png

**Comment by chyunsu3** [comment]
@avinashbarnwal Is it necessary to create a new branch? We can always revert changes later.

**Comment by avinashbarnwal** [comment]
Hi @hcho3,

I was thinking not to touch whatever is working. Please let me know what you think?

**Comment by chyunsu3** [comment]
@avinashbarnwal On the second thought, the second branch makes sense. We want to easily compare two implementations so that the new implementation can be validated against the current one. Go ahead and create new branch.

**Comment by avinashbarnwal** [comment]
Hi @hcho3 and Prof. @tdhock,

I have made the suggested changes in https://github.com/avinashbarnwal/xgboost/blob/survival_analysis1/src/objective/aft_obj.cc.

Please let me know your thoughts.

Now i need to change in the metric file. As we can change the loss as well left/right as interval.

I can create the class object AFTloss in the objective.cc and then call in metric survival_metric.cc.
But i am not sure how to do it?

Please let me know your thoughts.
Link of metric file :- https://github.com/avinashbarnwal/xgboost/blob/survival_analysis1/src/metric/survival_metric.cc

**Comment by chyunsu3** [comment]
@avinashbarnwal You need to mark functions `virtual` if you want child classes override these functions from the parent class.

**Comment by avinashbarnwal** [comment]
Yes. I was thinking the same. Thanks for the suggestions. I am making the changes now.

**Comment by tdhock** [comment]
for consistency of naming I would suggest something like "extreme value" rather than "weibull" because the other distributions you consider are normal and logistic, which are for real-valued numbers (the correspoonding positive valued distributions are called log-normal and log-logistic). Weibull is a positive-valued distribution whose real-valued counterpart is called the extreme value distribution.

**Comment by avinashbarnwal** [comment]
Yes. I will make the changes in the name straightaway. Please let me know if you think plots look good.

**Comment by tdhock** [comment]
why do you introduce the AFTEventType? seems overly complicated to me; would be simpler to just test if y_upper==y_lower then uncensored, otherwise censored. (and this logic can be coded only once in the AFTloss parent class)

again I don't think you should have four cases for computing the grad/hessian. instead just two cases (uncensored or interval) with interval having special cases for left/right censoring.

I would suggest moving grad_norm, hess_norm to members of the normal sub-class of the AFTloss class.

**Comment by avinashbarnwal** [comment]
Please check https://github.com/avinashbarnwal/xgboost/blob/survival_analysis1/src/objective/aft_obj.cc. We have only 2 cases. Event variable is important to categorize left/right and interval.

**Comment by tdhock** [comment]
name should be "extreme" to be consistent with R:
```r
> names(survival::survreg.distributions)
 [1] "extreme"     "logistic"    "gaussian"    "weibull"     "exponential"
 [6] "rayleigh"    "loggaussian" "lognormal"   "loglogistic" "t"          
> 
```

**Comment by tdhock** [comment]
the plots look reasonable, but the code is extremely unusual. `exp(1)**z` should be `exp(z)`

**Comment by tdhock** [comment]
I looked at aft_obj.cc but I don't see any reason to keep the event variable.

right now you have three cases, one for each censoring type, and each of them makes the same call, 
```
        first_order_grad  = aft::grad_interval(y_lower[i], y_higher[i], yhat[i],
                                              param_.aft_sigma,event);
```
I would suggest removing the event variable, removing the three cases, and simply call aft::grad_interval whenever there is censoring. Inside aft::grad_interval you can test if lower/upper limit is infinite.

**Comment by chyunsu3** [comment]
@tdhock We re-organized the code to incorporate your suggestion: https://github.com/avinashbarnwal/xgboost/commit/f52c96708ce54f12f594ec2a38537166b07d253f. We used polymorphism to massively simplify the metric and objective functions.

**Comment by tdhock** [comment]
ok but please further simplify https://github.com/avinashbarnwal/xgboost/commit/f52c96708ce54f12f594ec2a38537166b07d253f#r34598175

**Comment by avinashbarnwal** [comment]
I am looking into it.

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock and @hcho3,

I have pushed the extreme distribution to xgboost and also made changes to R files and plot with changed naming convention to "Extreme".

**Comment by tdhock** [comment]
how will that be specified in the user interface?

right now the "survival:cox" objective is documented on https://xgboost.readthedocs.io/en/latest/parameter.html as "negative values are considered right censored" so that implies to me that the "survival:" prefix means all outputs are positive-valued. (un-logged survival times)

to avoid confusion with that objective / output, I would suggest a new prefix for the three loss functions, "aft:gaussian" "aft:extreme" and "aft:logistic" -- does that make sense to you @hcho3  ? in each of these cases the output should be specfied by an interval (a,b) where both a and b are potentially negative. (log survival times) in other words, if you have an output vector `c(5.5, -0.1)` for survival:cox, that means the first output is un-censored, and the second is right censored (0.1, INFINITY) on the positive scale (survival times). The corresponding outputs that we would use are (1.7, 1.7) and (-2.3, INFINITY) --- on the real-valued scale (log survival times)

**Comment by chyunsu3** [comment]
@tdhock I think "survival:cox" uses survival time, not log survival time. So the output would be positive in that case. The current AFT implementation uses survival time as well.

**Comment by tdhock** [comment]
ok in that case the -0.1 output for survival:cox becomes (0.1, INFINITY) for us.
and in that case  "survival:lognormal"  "survival:loglogistic" "survival:weibull" may make more sense.

however I would suggest specfying outputs as real valued (potentially negative) log survival times, which has two advantages in my opinion
1. can use the entire double precision number range for storage (negative numbers can not be used if the output is positive / survival time)
2. consistency with the underlying real-valued distributions (gaussian, logistic, extreme)

maybe both aft:gaussian (using real valued log survival times) and survival:loggaussian (using positive valued un-logged survival times) etc can be supported?

**Comment by chyunsu3** [comment]
@tdhock That's reasonable suggestion. We will need to revise `survival:cox` implementation to use log survival times.

**Comment by avinashbarnwal** [comment]
Hi Prof @tdhock, @hcho3 

I am going through the document https://github.com/avinashbarnwal/GSOC-2019/blob/master/paper/AFT/survival.pdf shared by you. In page number 82, we have Extreme distribution section where it says 

"If y is Weibull then log(y) is distributed according to the (least) extreme
value distribution."

In previous cases - we have normal and logistic, where y is following normal and logistic.

Please let me know if i am missing anything.

I am trying to test extreme distribution using a notebook for the first few datasets. I am finding inf in errors. 
Link for notebook - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/test/data/neuroblastoma-data-master/src/notebook/002_xgboost_test.ipynb

ScreenShot of Gradient and hessian
![screenshot1](https://user-images.githubusercontent.com/6061417/62737420-bec5a880-b9fd-11e9-9eb4-7061f2b41b56.png)

I have used all the formulas mentioned in the document.

**Comment by tdhock** [comment]
the loss function for the extreme value dist goes to Inf very fast as the predicted value gets smaller.

if it helps, there is an alternative way to compute the loss (my.cost in code below) which is more numerically stable (results in Inf less often):

```
## extreme value CDF.
F <- function(z)1-exp(-exp(z))
max.pred <- 5
min.pred <- -15
curve(F, min.pred, max.pred)
s <- 1
y.lo <- -5
library(data.table)
pred.dt <- data.table(pred=seq(min.pred, max.pred, l=101))
pred.dt[, z.lo := (y.lo-pred)/s]
pred.dt[, p.lo := F(z.lo)]
pred.dt[, e.lo := 1/(exp(exp(z.lo))-1)]
pred.dt[, my.cost := log(1+e.lo)-log(e.lo)]
pred.dt[, naive.cost := -log(1-p.lo)]
plot(my.cost ~ pred, pred.dt)
pred.dt[, lines(pred, naive.cost, col="red")]
```

**Comment by avinashbarnwal** [comment]
Hi Prof @tdhock,

Please let me know what kind of censoring we have in the above formula? Looks like it is right censoring.

Do we have different stabilizing techniques for a different kind of censoring?

Please let me know what we are using finally for y in 3rd distribution extreme or Weibull?

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock and @hcho3 ,

I have added comment to https://github.com/avinashbarnwal/xgboost/commit/f52c96708ce54f12f594ec2a38537166b07d253f#r34598175.

**Comment by tdhock** [comment]
in my opinion the y values (at least internally) should be on the real-valued (potentially negative) scale, for all distributions, so the 3rd distribution should be "extreme"

that being said if you run into issues with extreme I think it would be more useful to just have a working prototype for the other two distributions.

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock and @hcho3,

I have kept "extreme" as 3rd distribution. I have changed the conditions to detect the censoring type as follows:-

1. Uncensored - Before -> y_lower == y_higher and After ->  y_lower == y_higher
2. Left - Before -> std::isinf(y_lower) and After -> std::abs(y_lower) <=1e-12f
3. Right - Before -> std::isinf(y_higher) and After -> std::isinf(y_higher)
4. Interval - Before -> !std::isinf(y_lower) && !std::isinf(y_higher) and After ->std::abs(y_lower)>1e-12f && !std::isinf(y_higher) 

Please let me if you think this is correct.

Rest of the code is same and I haven't made any changes to "Extreme" Distribution.

**Comment by tdhock** [comment]
i'm not sure that is correct.

y_lower and y_higher should be log survival times (possible negative).

if y_lower == y_higher then un-censored (use pdf)
otherwise censored 
- if y_lower == -INFINITY then p_lower=0 else p_lower = cdf(z_lower)
- if y_higher == INFINITY then p_higher=1 else p_higher = cdf(z_higher)
- loss = -log(p_higher - p_lower)

in particular I would suggest using INFINITY which is defined in math.h 

I don't think you should use the 1e-12 constant which is totally arbitrary.

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock,

We have been using y as un-logged values. Therefore, a comparison with 0 is done. I would change the y_lower and y_higher in logged format internally.

I saw that in your datasets you have log-transformed responses and to test I took anti-log of responses. 

Here, we expect users to enter survival times or logged transformed response. I am not sure about this?

**Comment by tdhock** [comment]
in survival::survreg the user can specify either kind of outputs. For me the logged outputs make more sense because anyway that is what we use internally.

**Comment by avinashbarnwal** [comment]
Thanks. I will change the y_lower and y_higher to log-transformed values and the conditions to detect the censoring types.

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock and @hcho3,

I have pushed the following changes:-

- Code refactoring for left/right/interval - loss,gradient and hessian.
- Removing x, mu, and sd and adding z.
- Now the conditions to detect censoring is same as before.
- Now we are passing log(y_lower) and log(y_higher) internally to do all the calculated. 

I have tested here - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/test/data/neuroblastoma-data-master/src/notebook/002_xgboost_test.ipynb. We have similar results as before.

**Comment by tdhock** [comment]
good. is there a PR where we can view a summary of all changes you want to merge? (files changed)

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock and @hcho3,

Please find the pull request - https://github.com/dmlc/xgboost/pull/4763.

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock and @hcho3 ,

I am trying to write the R vignette and i am finding this error. -

Error in setinfo.xgb.DMatrix(dtrain, "label_lower_bound", y_lower) : 
  setinfo: unknown info name label_lower_bound

Please find the R code - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/doc.R

Any help would be great.

**Comment by tdhock** [comment]
sounds like an xgboost error, @hcho3 should know

**Comment by tdhock** [comment]
is it normal in xgboost to specify the outputs/labels as attributes of the input/feature matrix? at least in R that is pretty confusing. the usual ways in R are
1. via formula, e.g. lm(y ~ x1 + x2, data=some.data.frame)
2. via separate input/output matrix objects e.g. glmnet(X, y) where X is the feature/input matrix and y is the output/label vector.

**Comment by chyunsu3** [comment]
@tdhock The mistake was due to a badly chosen `sigma` parameter. When I set `sigma` to 10 or 20, the metrics are no longer infinity. In general, we should choose a reasonable value for `sigma` depending on the range of log survival times. The particular dataset has log survival time ranging from 8 to 10, so `sigma=1` is not reasonable.

**Comment by chyunsu3** [comment]
@tdhock I do not think XGBoost package supports the use of a formula. Currently, users are asked to create an object of type `xgb.DMatrix` with data and labels supplied.

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock,

@hcho3 helped in resolving the issue. We have a working vignette here - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/R/doc.R. I am structuring more.

**Comment by tdhock** [comment]
red flag!

why doesn't sigma=1 work? I think it should be fine for the data sets I have provided.

sigma=10 seems too big to me -- there is almost no flat region for interval-censored outputs.

```r
s <- 1;curve(-log(pnorm(10,x,s)-pnorm(5,x,s)),0, 20)
```

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock,

I think we need to see the standard deviation of log(survival times) and we might use something similar sigma as starting point. We tested it with very less sigma and it was giving very less predicted value with very high log(survival times). For example - log(survival times) ~ 10 and y hat was 0.03. I think this might not be correct. 

Please let me know your thoughts.

**Comment by chyunsu3** [comment]
@tdhock @avinashbarnwal Maybe we need set a bigger value for `base_score`, the initial guess for yhat. By default, it is set to 0.5, which may be too low. Let's try running XGBoost with `sigma=1` and `base_score=8`.

**Comment by tdhock** [comment]
more generally you could set it to the mean of all the finite limits?

**Comment by avinashbarnwal** [comment]
Hi Prof. @tdhock and @hcho3,

I have included base_score as average and sigma as the standard deviation of the log(survival_times).
We have very less number of Infinite observations now.

NoteBook - https://github.com/avinashbarnwal/GSOC-2019/blob/master/AFT/test/data/neuroblastoma-data-master/src/notebook/002_xgboost_test.ipynb
